hashing/src/chunk_with_proof.rs

```diff
@@ -0,0 +1,307 @@
+impl ToBytes for ChunkWithProof {
+    fn to_bytes(&self) -> Result<Vec<u8>, casper_types::bytesrepr::Error> {
+        let mut result = allocate_buffer(self)?;
+        result.append(&mut self.proof.to_bytes()?);
+        result.append(&mut Bytes::from(self.chunk.as_slice()).to_bytes()?);
+        Ok(result)
+    }
+
+    fn serialized_length(&self) -> usize {
+        self.proof.serialized_length() + self.chunk.serialized_length()
+    }
+}
+
+impl FromBytes for ChunkWithProof {
+    fn from_bytes(bytes: &[u8]) -> Result<(Self, &[u8]), casper_types::bytesrepr::Error> {
+        let ((proof, chunk), remainder) = <(IndexedMerkleProof, Bytes)>::from_bytes(bytes)?;
+
+        Ok((
+            ChunkWithProof::try_from(ChunkWithProofDeserializeValidator {
+                proof,
+                chunk: chunk.into(),
+            })
+            .map_err(|_| casper_types::bytesrepr::Error::Validation)?,
+            remainder,
+        ))
+    }
+}
+
+impl ChunkWithProof {
+    #[cfg(test)]
+    pub const CHUNK_SIZE: usize = 10;
+
+    #[cfg(not(test))]
+    pub const CHUNK_SIZE: usize = 1_048_576; // 2^20
+
+    #[cfg(test)]
+    pub fn new(data: &[u8], index: u64) -> Result<Self, error::MerkleConstructionError> {
+        if data.len() < Self::CHUNK_SIZE * (index as usize) {
+            return Err(error::MerkleConstructionError::IndexOutOfBounds {
+                count: data.chunks(Self::CHUNK_SIZE).len() as u64,
+                index,
+            });
+        }
+
+        let (proof, chunk) = if data.is_empty() {
+            (IndexedMerkleProof::new([Digest::hash(&[])], index)?, vec![])
+        } else {
+            (
+                IndexedMerkleProof::new(data.chunks(Self::CHUNK_SIZE).map(Digest::hash), index)?,
+                data[Self::CHUNK_SIZE * (index as usize)
+                    ..data.len().min(Self::CHUNK_SIZE * ((index as usize) + 1))]
+                    .to_vec(),
+            )
+        };
+
+        Ok(ChunkWithProof { proof, chunk })
+    }
+
+    fn is_valid(&self) -> bool {
+        let chunk_hash = Digest::hash(self.chunk());
+        self.proof
+            .merkle_proof()
+            .first()
+            .map_or(false, |first_hash| chunk_hash == *first_hash)
+    }
+
+    /// Get a reference to the chunk with proof's chunk.
+    pub fn chunk(&self) -> &[u8] {
+        self.chunk.as_slice()
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use crate::{error, Digest};
+    use casper_types::bytesrepr::{FromBytes, ToBytes};
+    use proptest::proptest;
+    use rand::Rng;
+    use std::convert::TryInto;
+
+    use crate::chunk_with_proof::ChunkWithProof;
+
+    proptest! {
+    #[test]
+    fn generates_valid_proof(chunk_count in 0usize..32usize) {
+        let data = vec![0u8; ChunkWithProof::CHUNK_SIZE * chunk_count];
+
+        let number_of_chunks: u64 = data
+            .chunks(ChunkWithProof::CHUNK_SIZE)
+            .len()
+            .try_into()
+            .unwrap();
+
+        assert!(!(0..number_of_chunks)
```

Comment:
`rustfmt` and `clippy` fail here... I think we should make a `ChunkCount(usize)` newtype with an `Arbitrary` impl which makes it between 0 and 32.  Then you can use the `proptest` proc-macro.

Reply:
Done, with a caveat described in my other comment.